### ISSUE TITLE
AssertionError: graph-captured input #1, of type <class 'torch.Tensor'>, is not among original inputs of types…

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#112043]
I know we have iterated a bunch on this error message, but still: putting on my user hat, I find this to be really inscrutable.

1. What is a "graph-captured input"? What causes an input to be "graph-captured"?
2. Which input is input 1? What determines the ordering of the inputs?
3. What would cause the graph-captured inputs to be different from the the original inputs?
4. etc.

It's definitely really challenging to come up with a good explanation for wtf is happening here since it's so dependent on implementation details. Ideally we can make it so that this error just…never happens somehow.

cc @ezyang @gchanan @zou3519 @kadeng @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi who I know has been thinking about this recently.

**Comment by tugsbayasgalan** [comment]
Yeah we plan to address this. I think the correct fix should be that export knows how to handle a graph coming from dynamo that has potentially different input/output than user calling convention.

**Comment by suo** [comment]
Here is a subset of the failure domain we could chunk off and make clearer:
```
from dataclasses import dataclass

import torch
from torch import Tensor


@dataclass
class Input:
    foo: Tensor


class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, input: Input):
        return input.foo + input.foo


ep = torch.export.export(Mod(), (Input(foo=torch.ones(2, 3)),))
```

produces
```
AssertionError: graph-captured input #1, of type <class 'torch.Tensor'>, is not among original inputs of types: (<class '__main__.Input'>)
```

This is because we didn't register a pytree node for `Input`.

Probably we should check that all orig inputs have a corresponding pytree node (or are leaves) and error out early if so. @tugsbayasgalan would you prefer I file a separate issue for this or keep it here?

**Comment by tugsbayasgalan** [comment]
Option 1 is preferred

**Comment by ezyang** [comment]
I would be more receptive to the idea of point fixing our way to victory if the point fixes were simple to add. Empirically, bug fixes for these problems rarely ever actually show up, at least partially because no one on the export team feels that monkeying around with the Dynamo logic that does the matching.

I would propose an option 3, where we rewrite the Dynamo logic. Instead of trying to reverse engineer inputs using tensor id matching and pytree traversal, we instead use Dynamo's concept of Source for the source of truth for performing the correspondence. Every input to the inner FX graph is associated with a source that looks something like `L['foo']` or `L['self'].buffer_foo`. Some sources are permissible as export inputs as they correspond to obvious inputs to the function in question, while others (like NN module sources) are not. This will permit you to give a good error message specifying exactly what access happened, then the only other sugar you need is remembering where in user code this access actually happened so you can fix the model.
